internal/pkg/agent/storage/store/migrations.go

```diff
@@ -0,0 +1,204 @@
+		if err == nil {
+			err = actionDiskStore.Delete()
+			if err != nil {
+				log.Errorf("failed to delete action store %s after migration: %v", actionStorePath, err)
+			}
+		}
+	}()
+
+	action, err := migrations.LoadActionStore(actionDiskStore)
+	if err != nil {
+		log.Errorf("failed to load action store for migration %s: %v",
+			actionStorePath, err)
+		return err
+	}
+
+	// no actions stored nothing to migrate
+	if action == nil {
+		log.Debugf("no action stored in the action store %s, nothing to migrate", actionStorePath)
+		return nil
+	}
+
+	if action.Type != fleetapi.ActionTypePolicyChange {
+		log.Warnf("unexpected action type when migrating from action store. "+
+			"Found %s, but only %s is suported. Ignoring action and proceeding.",
+			action.Type, fleetapi.ActionTypePolicyChange)
+		// If it isn't ignored, the agent will be stuck here and require manual
+		// intervention to fix the store.
+		return nil
+	}
+
+	stateStore, err := NewStateStore(log, stateDiskStore)
+	if err != nil {
+		return err
+	}
+
+	// set actions from the action store to the state store
+	stateStore.SetAction(&fleetapi.ActionPolicyChange{
+		ActionID:   action.ActionID,
+		ActionType: action.Type,
+		Data: fleetapi.ActionPolicyChangeData{
+			Policy: conv.YAMLMapToJSONMap(action.Policy),
+		},
+	})
+
+	err = stateStore.Save()
+	if err != nil {
+		log.Debugf("failed to save agent state store, err: %v", err)
+	}
+	return err
+}
+
+// migrateYAMLStateStoreToStateStoreV1 migrates the YAML store to the new JSON
+// state store. If the contents of store is already a JSON, it returns the
+// parsed JSON.
+func migrateYAMLStateStoreToStateStoreV1(log *logger.Logger, store storage.Storage) error {
+	exists, err := store.Exists()
+	if err != nil {
+		return fmt.Errorf("migration YMAL to Store v1 failed: "+
+			"could not load store from disk: %w", err)
+	}
+
+	// nothing to migrate, return empty store.
+	if !exists {
+		return nil
+	}
+
+	// JSON is a subset of YAML, so first check if it's already a JSON store.
+	reader, err := store.Load()
+	if err != nil {
+		return fmt.Errorf("could not read store content: %w", err)
+	}
+
+	st, err := readState(reader)
+	// close it as soon as possible and before the next store save
+	_ = reader.Close()
+	if err == nil {
+		// it's a valid JSON, therefore nothing to migrate
+		return nil
+	}
+
+	// Try to read the store as YAML
+	yamlStore, err := migrations.LoadYAMLStateStore(store)
+	if err != nil {
+		// it isn't a YAML store
+		return errors.Join(ErrInvalidYAML, err)
+	}
+
+	// Store was empty, nothing to migrate
+	if yamlStore == nil {
+		return nil
+	}
+
+	var action fleetapi.Action
+	switch yamlStore.Action.Type {
+	case fleetapi.ActionTypePolicyChange:
+		action = &fleetapi.ActionPolicyChange{
+			ActionID:   yamlStore.Action.ActionID,
+			ActionType: yamlStore.Action.Type,
+			Data: fleetapi.ActionPolicyChangeData{
+				Policy: conv.YAMLMapToJSONMap(yamlStore.Action.Policy)},
+		}
+	case fleetapi.ActionTypeUnenroll:
```

Comment:
Similar to above, how can we help people avoid forgetting to add migrations if they extend the store with a new type sometime later?

I do not want this problem to surface only after release.

Reply:
What I am worried about is we released 8.16.0 with a new action type that hasn't occurred yet so we don't migrate it, then we release 8.16.1 and at the time of upgrade the 8.16.0 agent has the action in the store, but 8.16.1 doesn't know how to handle it.

You are right that this doesn't matter for the first upgrade where old agents can't send the new action yet, but it will matter for every upgrade from that version onward.

Reply:
I suppose in this situation we will already be the YAML store so we don't need to do a migration at all and everything is fine.

Reply:
> What I am worried about is we released 8.16.0 with a new action type that hasn't occurred yet so we don't migrate it [...]

This won't happen, as soon as this PR is released, there will be no mo migration needed. Unless another refactor is done on the state store. Once it's released, if we add a new action type to be stored, it'll be added on the new store. Besides now there is a version on the store, making it easier to migrate or evolve the store.

Reply:
for that to happen, we'd need to release a new action being persisted, then after that have this PR merged without updating it. Which I'd say has a negligible chance of happening, besides most likely changing the store will always result on a merge conflict with this PR, forcing it to be updated.